docs/efficient_gas_calculation_algorithm.md

```diff
@@ -0,0 +1,156 @@
+# Efficient gas calculation algorithm for EVM
+
+This article describes how to efficiently calculate gas and check stack requirements
+for Ethereum Virtual Machine (EVM) instructions.
+
+
+## Instructions metadata
+
+Let's start by defining some basic and universal instructions' parameters.
+
+1. Base gas cost.
+   
+   This is the static gas cost of instructions. Some instructions may have 
+   additional cost depending on their operand values - these have to be
+   handled individually during the instruction execution.
+   
+2. Stack height requirement.
+
+   This is the minimum stack height (number of items on the stack) 
+   required for the instruction execution.
+   
+3. Stack height change.
+
+   This is difference of the stack height before and after the instruction 
+   execution. Can be negative if the instruction pops more items than pushes.
+   
+Examples:
+
+| opcode  | base gas cost | stack height requirement | stack height change |
+| ------- | ------------- | ------------------------ | ------------------- |
+| ADD     | 3             | 2                        | -1                  |
+| EXP     | 50            | 2                        | -1                  |
+| DUP4    | 3             | 4                        | 1                   |
+| SWAP1   | 3             | 2                        | 0                   |
+| ADDRESS | 2             | 0                        | 1                   |
+| CALL    | 700           | 7                        | -6                  |
+
+
+## Basic blocks
+
+A _basic block_ is a sequence of instructions that are executed "straight-line"
+without being interrupted by jumps. I.e. they are nodes in the _control flow graph_.
+
+The name "basic block" has been taken from LLVM. Other names are: just "block"
+in wasm or "subroutine" in subroutine-threaded interpreters.
+
+In EVM there are simple rules to identify basic block boundaries:
+
+1. The following instructions _end_ a basic block:
+   - `JUMP`,
+   - `JUMPI`,
+   - `STOP`,
+   - `RETURN`,
+   - `REVERT`,
+   - `SELFDESTRUCT`,
+   - an instruction directly preceding `JUMPDEST`.
+
+2. The following instructions _start_ a basic block:
+   - `JUMPDEST`,
```

Comment:
```suggestion
```
(this should be redundant together with the two rules below)

Reply:
I don't believe this is true. You can have a `JUMPDEST` anywhere.

Reply:
sure, but you say: a basic block starts right before ... an instruction directly after an intruction that has ended the previous basic block. And above: a basic block ends after ... an instruction directly preceding "jumpdest". Maybe just nitpicking :)

Reply:
I'm wondering about the attackability of this -- in hte case where a contract consists of 24k `JUMPDEST`s. I guess that's up to the implementation.

Reply:
I have a test/benchmark for this one. Other variations about it are definitely needed.

evmone's top speed is ~1Ggas/s. In this case where we do gas/stack checking for every instruction the execution is ~0.5Ggas/s.

The bigger problem is the analysis of this code takes ~0.1 ms what means it should cost ~1000 gas. This is the biggest concern so far.